K8petstore

jayunit100 opened the pull request:
Here it is ! I've finished automating and testing this at 20 and 40 nodes, and added a throughput indicator.  

You can see some interesting behaviour happens at times, i.e. flatlining, haven't determined quite exactly why yet, could be a minor glitch in the generator , but since its stochastic, i think its related to stability of containers in certain scenarios.  

Ive ran this on up to 40 nodes to generate up to around 500,000 transactions in total.  

Would love to scale it further.

All docker code is contained here, including the redis containers themselves so things like /etc/redis conf files can be customized, and also, so that we pegs a specific version to avoid goredis api compatibility syncing issues...  w/ excpetion of the bigpetstore data generator, which is already curated in apache bigtop (documentation is in the README for that as well).

To run ?  Just run `k8petstore.sh` :)

jayunit100 committed:
K8PetStore: A scalable kubernetes application with automated load generation (#3137).

satnam6502 commented:
@jayunit100 please can you suggest a reviewer(s)?

jayunit100 commented:
Hi satnam.  Possbly @brendanburns , and yourself @satnam6502 might be able to review this .   It should be very easy to test (just run `k8petstore.sh` and modify the variables at the top, which are self explanatory, as you see fit)

jayunit100 commented:
(fyi, the official issue tracker is https://github.com/GoogleCloudPlatform/kubernetes/issues/5260 )

satnam6502 commented:
Thanks for the suggestions. I'm on call today so I've asked Brendan to help out.

brendandburns commented on examples/k8petstore/README.md:
```diff
@@ -0,0 +1,123 @@
+
+## Architecture of this SOA
+
+A diagram of the overall architecture of this application can be seen in arch.dot (you can paste the contents in any graphviz viewer, including online ones such as http://sandbox.kidstrythisathome.com/erdos/.
```
linkify arch.dot ?

brendandburns commented on examples/k8petstore/README.md:
```diff
@@ -0,0 +1,123 @@
+
+If you are new to kubernetes, and you haven't run guestbook yet, 
+
+you might want to stop here and go back and run guestbook app first.  
```
link to the guestbook app readme?

brendandburns commented on examples/k8petstore/README.md:
```diff
@@ -0,0 +1,123 @@
+
+like the number of data generators (more generators will create more load on the redis master).
+
+So, to run this app in kubernetes, simply run `k8petstore.sh`.
```
linkify k8petstore.sh?

brendandburns commented on examples/k8petstore/README.md:
```diff
@@ -0,0 +1,123 @@
+
+## Questions
+
+For questions on running this app, you can ask on the google containers group.
```
google-containers@googlegroups.com or #google-containers on IRC.

brendandburns commented:
Cool, some minor comments on the docs, but generally LGTM.

Thanks for doing this!
--brendan

jayunit100 committed:
Cleanup after initial PR and shading of src.

jayunit100 commented:
okay updated @brendanburns  !

brendandburns commented:
Sorry, one more thing, can you run gofmt on:

```./examples/k8petstore/web-server/src/main/PetStoreBook.go```

Thanks!
--brendan

jayunit100 commented:
okay, done.  thats a neat tool.  didn't know about it

brendandburns commented:
Ugh, travis is trying to build your example code:

https://travis-ci.org/GoogleCloudPlatform/kubernetes/jobs/54292082

We should either include the right libraries in our Gopath, or block travis from building this somehow.

jayunit100 commented:
hiya, okay, i;ll look into adding the libraries.  Im not sure where to add them so if its non obvious just ping in this thread :)... will update when im done.

jayunit100 commented:
okay, pushed an to GoDeps ........... we will see if it works.  thanks for noticing that !~

jayunit100 commented:
Well, looks like the build still fails.  I think i have a way to reproduce faithfully now .  I guess one option is to rename `src` to `_src` .... but that is only a temporary fix.

jayunit100 commented:
Well, I got the build working locally: 
```
[jay@rhbd kubernetes]$  go build -v ./... 2>&1 | grep examples
_/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/examples
_/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/examples/liveness/image
_/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/examples/k8petstore/web-server/src/main
[jay@rhbd kubernetes]$ pwd
/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes
```
I will see if i can follow the directions in the developer docs to build it cleanly.  maybe its just working b.c of the local deps which are on my machine, (i.e. possibly i didn't add them correctly into the Godeps json file)

jayunit100 commented:
@brendandburns okay, i did the temporary disablement of the `src` dir by moving to `_src`.  

As a follow on , lets work on https://github.com/GoogleCloudPlatform/kubernetes/issues/5465, as part of which we can get rid of needing to do stuff like that, and bring `examples/` into the build

jayunit100 commented:
(bump) tl;dr this is now passing and ready to merge - i shaded the src package to `_src`, and updated the dependencies.  maybe in a future build, it will pass once travis integrates examples/ src properly.

smarterclayton commented:
Squash before merge (once final sign offs are given) please.

jayunit100 commented:
Squashed into 2 commits (original patch in the first ; cleanup in the second).

brendandburns commented:
LGTM, will merge on green.

j3ffml committed:
Add dependencies for k8petstore to Godeps
Cleanup k8petstore to use Godeps dependencies, add to hack/build-go.sh
Remove trailing spaces, README

jlowdermilk commented:
lgtm, but needs rebase and squash.

jayunit100 commented:
okay , rebased and squashed into 3 commits (one containing the fixes which jeff pushed, one containing the cleanups which i made after the original patch, and one with the original code dump).   thanks for your help Jeff and Brendan !

brendandburns commented:
merged!
